account_financial_report_webkit/models/account_static_balance.py

```diff
@@ -53,119 +61,179 @@ def unlink(self):
+        :param period: period to calculate the values for
         :param accounts: optional accounts to calculate the values for
+        :param journals: optional journals to calculate the values for
         :return: True
         """
-        if not periods:
+        if not period:
             return {}
-        if accounts is None:
-            # Calculate for all accounts
-            accounts = self.env['account.account'].search([])
+
+        # Fetch accounts used in this period if no accounts are given
+        if not accounts:
+            query = """
+                SELECT acc.id FROM account_account acc
+                WHERE exists( SELECT FROM account_move_line aml
+                    WHERE period_id = %(period_id)s and account_id = acc.id )
+                    """
+            self.env.cr.execute(query, {
+                'period_id': period.id
+            })
+            account_ids = self.env.cr.fetchall()
+            accounts = self.env['account.account'].search([
+                ('id', 'in', account_ids)
+            ])
+
+            if not accounts:
+                logger.debug(
+                    'Calculation of balance for period %s is cancelled, '
+                    'because there are no accounts active', period.name)
+                return {}
+
+        # Fetch journals used in this period if no journals are given
+        if not journals:
+            query = """
+                SELECT aj.id FROM account_journal aj
+                WHERE exists( SELECT FROM account_move_line aml
+                    WHERE period_id = %(period_id)s and journal_id = aj.id )
+                    """
+            self.env.cr.execute(query, {
+                'period_id': period.id
+            })
+            journal_ids = self.env.cr.fetchall()
+            journals = self.env['account.journal'].search([
+                ('id', 'in', journal_ids)
+            ])
+
+            if not journals:
+                logger.debug(
+                    'Calculation of balance for period %s is cancelled, '
+                    'because there are no accounts active', period.name)
+                return {}
 
         logger.debug(
             "Starting calculation of balances for "
-            "periods: %s.", ", ".join(periods.mapped('name')))
+            "periods: %s.", period.name)
 
         start_time = time()
 
+        for journal in journals:
+            balances = self.env['account.static.balance'].search([
+                ('period_id', '=', period.id),
+                ('journal_id', '=', journal.id)
+            ])
+            existing_accounts = balances.mapped('account_id')
+
+            if existing_accounts and existing_accounts != accounts:
+                balances.unlink()
+
+        # Filter out journals that are already calculated
+        existing_journals = self.env['account.static.balance'].search([
+            ('period_id', '=', period.id)
+        ]).mapped('journal_id')
```

Comment:
I still don't agree with this flow in which you fetch all missing combinations, reduce them to periods and then determine the missing journals from the existing journal-period combinations.

Consider the following: we have given the Finance department deletion rights on account.static.balance so that if they (or 'we') see that anything is wrong with the balances for whatever reason they can delete some of them to trigger a recompute of those account/period combinations. Due to this new logic, the recompute may not happen. Instead, more inconsistencies are introduced in the reports. Now, I don't know if that is a well designed flow or if it is actually used, but I'm thinking that if you have the actual information of the missing journals per period in the calling method you might as well use it.

Reply:
For the flow you described, we have the scheduled action `auto_populate_table` that will recompute the new missing balances.

Reply:
Dived a little bit deeper into it. If someone unlinks the account.static.balance, there is indeed a trigger to recalculate the balances: https://github.com/EssentNovaTeam/account-financial-reporting/pull/6/files#diff-c4749f6b2ebab521a3de0b3add26b25c49bc78e8c8a4271f430e91f2cf6b4f54R60-R70

But because of the unlink (of one account journal combination) this filter wont succeed: https://github.com/EssentNovaTeam/account-financial-reporting/pull/6/files#diff-c4749f6b2ebab521a3de0b3add26b25c49bc78e8c8a4271f430e91f2cf6b4f54R145-R152

Next step is that everything for this journal period will be unlinked, and won't be available anymore as an existing_journal.